### PR TITLE
Module optimization

### PR DESCRIPTION
**Pull request by AsaNehm**
This script allows for plotting and evaluation of parameters for the geometry optimization.

These parameters are (can be called each on their own):
- Start point resolution
- End point resolution
- Charge ID performance (total and per KE slices)
- Angular resolution (total and per KE slices)

### CODE REVIEW

File: scripts/Reco/optimization_evaluation.py [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+            if plot_Angle or plot_Charge:
+                Reco_Track_StartDirection = np.frombuffer(event.StartDirection, dtype = np.float32)
+                MomentumTrackStart = np.frombuffer(true_event.RecoTrackPrimaryParticleTrueMomentumTrackStart, dtype = np.float32)
+            # charge related stuff
+            if plot_Charge:
+                Reco_Track_Charge = event.Charge
+            True_PDG = true_event.PDG
+            # classify if true muon actually enough distance in TMS and start in LAr
+            LArFiducialTouch = true_event.RecoTrackPrimaryParticleLArFiducialStart
+            Particle_PDG = true_event.LeptonPDG
```
Comment (jdkio) [inline scripts/Reco/optimization_evaluation.py:124]: Use `RecoTrackPrimaryParticlePDG` to find the PDG of the reco track specifically. `LeptonPDG` ends up being the PDG of the lepton of the primary event, where the primary event is defined as the event that deposits the most visible energy. Usually these are one and the same but there's no guarantee

File: scripts/Reco/optimization_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+            Muon_Start = np.frombuffer(true_event.Muon_Vertex, dtype = np.float32)
+            Muon_End = np.frombuffer(true_event.Muon_Death, dtype = np.float32)
```
Comment (jdkio) [inline scripts/Reco/optimization_evaluation.py:126]: Similar to `LeptonPDG`, it's better to use `RecoTrackPrimaryParticleTruePositionStart` and `RecoTrackPrimaryParticleTruePositionEnd`

File: scripts/Reco/optimization_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+            True_Position_TMS_Start = np.frombuffer(true_event.RecoTrackPrimaryParticleTruePositionEnteringTMS, dtype = np.float32)
+            EndPos = np.frombuffer(event.EndPos, dtype = np.float32)
+            True_Position_TMS_End = np.frombuffer(true_event.RecoTrackPrimaryParticleTruePositionLeavingTMS, dtype = np.float32)
+            # track direction/momentum at start
+            if plot_Angle or plot_Charge:
+                Reco_Track_StartDirection = np.frombuffer(event.StartDirection, dtype = np.float32)
+                MomentumTrackStart = np.frombuffer(true_event.RecoTrackPrimaryParticleTrueMomentumTrackStart, dtype = np.float32)
+            # charge related stuff
+            if plot_Charge:
+                Reco_Track_Charge = event.Charge
```
Comment (jdkio) [inline scripts/Reco/optimization_evaluation.py:120]: Does this need a `np.frombuffer(event.Charge, dtype = np.float32)`

File: scripts/Reco/optimization_evaluation.py (hunk 1 above)
Comment (AsaNehm) [reply scripts/Reco/optimization_evaluation.py:124]: Ok. Will change this

File: scripts/Reco/optimization_evaluation.py (hunk 2 above)
Comment (AsaNehm) [reply scripts/Reco/optimization_evaluation.py:126]: The intention for these two are to check whether the muon actually starts in ND-LAr and ends in TMS. For the latter I could definitely switch over to the Reco... one though

File: scripts/Reco/optimization_evaluation.py (hunk 3 above)
Comment (AsaNehm) [reply scripts/Reco/optimization_evaluation.py:120]: As it works it doesn't. I guess that single numbers don't need the frombuffer one, while array-like things need it

**Comment by AsaNehm** [comment]
Added some changes to the event display (draw_spill_3D_projections.py) to fix the overwriting of events in the same spill and added a legend as well

File: scripts/Reco/optimization_evaluation.py (hunk 2 above)
Comment (jdkio) [reply scripts/Reco/optimization_evaluation.py:126]: In theory, you could have two muons in an event, so it's better to look at the actual muon that was reconstructed. That's what the `RecoTrackPrimaryParticle*` variables are for

**Review by LiamOS — APPROVED** [review]

**Pull request merged**